crowbar_framework/lib/crowbar/lock.rb

```diff
@@ -0,0 +1,76 @@
+#
+# Copyright 2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  class Lock
+    attr_accessor :logger
+    attr_accessor :path
+    attr_accessor :remote
+
+    def initialize(options = {})
+      @logger = options.fetch :logger, Rails.logger
+      @path = options.fetch :path, "/var/lock"
+      @remote = options.fetch :remote, false
+    end
+
+    def create(name, nodes = [])
+      if @remote
+        remote(name, nodes, action: "create")
+      else
+        local(name, action: "create")
```

Comment:
[We already have local locking](https://github.com/crowbar/barclamp-crowbar/blob/master/crowbar_framework/lib/file_lock.rb), and it actually works, unlike this implementation which is completely and utterly broken with respect to concurrency.  In fact it would be better to reuse that implementation by making it available on the other nodes and then keeping this class as a wrapper which executes it remotely via `ssh_cmd`.  Alternatively `ssh_cmd` could run `flock(1)` on the remote node.  The other suggestion I already made is to use `symlink(2)` or `mkdir(2)` which are both atomic operations (unlike just writing a file).